lib/Dialect/FIRRTL/FieldRef.cpp

```diff
@@ -0,0 +1,127 @@
+#include "circt/Dialect/FIRRTL/FieldRef.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace firrtl;
+
+Operation *FieldRef::getDefiningOp() const {
+  if (auto *op = value.getDefiningOp())
+    return op;
+  return value.cast<BlockArgument>().getOwner()->getParentOp();
+}
+
+FieldRef::FieldRef(Value val) : value(val) {
+  // This code walks upwards from the subfield and calculates the field ID at
+  // each level. At each stage, it must take the current id, and re-index it as
+  // a nested bundle under the parent field.. This is accomplished by using the
+  // parent field's ID as a base, and adding the field ID of the child.
+  //
+  // a (0) { b (1) c (2) }
+  while (val) {
+    Operation *op = val.getDefiningOp();
+
+    // If this is a block argument, we are done.
+    if (!op)
+      return;
+
+    // TODO: implement subindex op.
+    TypeSwitch<Operation *>(op)
+        .Case<SubfieldOp>([&](auto op) {
+          auto bundleType = op.input().getType().template cast<BundleType>();
+          auto index = bundleType.getElementIndex(op.fieldname()).getValue();
+          // Rebase the current index on the parent field's index.
+          id = bundleType.getFieldID(index) + id;
+          val = op.input();
+        })
+        .Default([&](auto op) {
+          // We are done walking the chain.
+          value = val;
+          val = nullptr;
+        });
+  }
+}
+
+/// Get the string name of a value which is a direct child of a declaration op.
+static void getDeclName(Value value, SmallString<64> &string) {
+  if (auto arg = value.dyn_cast<BlockArgument>()) {
+    // Get the module ports and get the name.
+    auto module = cast<FModuleOp>(arg.getOwner()->getParentOp());
+    SmallVector<ModulePortInfo> ports = module.getPorts();
+    string += ports[arg.getArgNumber()].name.getValue();
+    return;
+  }
+
+  auto *op = value.getDefiningOp();
+  TypeSwitch<Operation *>(op)
+      .Case<InstanceOp, MemOp>([&](auto op) {
+        string += op.name();
+        string += ".";
+        string +=
+            op.getPortName(value.cast<OpResult>().getResultNumber()).getValue();
+      })
+      .Case<WireOp, RegOp, RegResetOp>([&](auto op) { string += op.name(); });
+}
+
+std::string FieldRef::getFieldName() const {
+  // This will find the index of the bundle element with a field ID which
+  // contains the searched for ID.  This means that it returns the index of the
+  // greatest less-than field ID.
+  auto findIndexForFieldID = [](BundleType bundleType, unsigned id) {
+    assert(bundleType.getElements().size() && "Bundle must have >0 fields");
+    // Find the field corresponding to this element using a binary search.
+    unsigned l = 0;
+    unsigned r = bundleType.getElements().size() - 1;
+    while (l < r) {
+      auto m = l + (r - l + 1) / 2;
+      if (id < bundleType.getFieldID(m)) {
+        r = m - 1;
+      } else {
+        l = m;
+      }
+    }
+    return l;
+  };
```

Comment:
small explanation about why I wrote my own binary search: the std algorithms didn't quite work for my use case:
```yaml
std::lower_bound: first element that is not less than (i.e. greater or equal to) value,
std::upper_bound: first element that is greater than value
findIndexForFieldID: last element that is less than the value
```
I might have been able to get this to work with an inverted comparator and std::upperbound or something? I didn't think of this before but I'm inclined to leave this as is unless some one has a preference

Reply:
std::advance(lower-bound (), -1)?

Reply:
Thanks!

Reply:
Fixed! I moved this logic in to `BundleType` which I think cleaned this up a bit.